### DISCUSSION TITLE
Investigate why NATS subscribe lower than Kafka in Watermill benchmark?

### DISCUSSION THREAD

**Discussion by kokizzu** [nats-io/nats-server#5222] [General]
- [x] Verification Request

https://github.com/ThreeDotsLabs/watermill#benchmarks

**Comment by derekcollison** [comment]
Will take a look, not familiar with that one.

**Comment by kozlovic** [comment]
Not sure if this is the same, but I did post some comments back in September 2019: https://github.com/ThreeDotsLabs/watermill-benchmark/issues/2

This was for NATS Streaming though, not sure if the issue here is for core NATS or not.

**Comment by matthiashanel** [comment]
They are still using mats streaming. With v0.11.2

**Comment by derekcollison** [comment]
Can we make any updates here?

**Comment by AlexCuse** [comment]
Recently introduced a rough cut of v2 for https://github.com/ThreeDotsLabs/watermill-nats with Jetstream and experimental NATS core support.  Hoping to find some time to update the benchmark project accordingly.

**Comment by AlexCuse** [comment]
The slowness was likely coming from use of `GobMarshaler` to encode the watermill message by default in the old implementation.  New implementation conveys the watermill message details in headers by default and should bring almost no additional marshaling / unmarshaling overhead beyond app payload handling.  Double encoding like that is always a killer especially w big messages.

**Comment by AlexCuse** [comment]
Have a PR open to update to new NATS middleware, see about 45k msg/sec throughput in the default configuration and 120k with asynchronous acks enabled. 

Will see about adding a nats-core benchmark next but that's a little tricky since existing stuff is written to fill a stream and then drain it.  See about 90k msg/sec publishing.  Running 1 sub per core on an 8 core laptop.  If there's a reference docker-compose for setting up a cluster I can try that as well, though with the broker running on the same host as the subscribers I'm not sure how interesting that is.
